backends/backends.go

```diff
@@ -0,0 +1,476 @@
+	var authenticated bool
+	var err error
+
+	// If prefixes are enabled, check if username has a valid prefix and use the correct backend if so.
+	if !b.checkPrefix {
+		return b.checkAuth(username, password, clientid)
+	}
+
+	validPrefix, bename := b.lookupPrefix(username)
+
+	if !checkRegistered(bename, b.userCheckers) {
+		return false, fmt.Errorf("backend %s not registered to check users", bename)
+	}
+
+	if !validPrefix {
+		return b.checkAuth(username, password, clientid)
+	}
+
+	// If the backend is JWT and the token was prefixed, then strip the token. If the token was passed without a prefix it will be handled in the common case.
+	if bename == jwtBackend {
+		prefix := b.getPrefixForBackend(bename)
+		username = strings.TrimPrefix(username, prefix+"_")
+	}
+	var backend = b.backends[bename]
+
+	authenticated, err = backend.GetUser(username, password, clientid)
+	if authenticated && err == nil {
+		log.Debugf("user %s authenticated with backend %s", username, backend.GetName())
+	}
+
+	return authenticated, err
+}
+
+func (b *Backends) checkAuth(username, password, clientid string) (bool, error) {
+	var err error
+	authenticated := false
+
+	for _, bename := range b.userCheckers {
+		var backend = b.backends[bename]
+
+		log.Debugf("checking user %s with backend %s", username, backend.GetName())
+
+		if ok, getUserErr := backend.GetUser(username, password, clientid); ok && getUserErr == nil {
+			authenticated = true
+			log.Debugf("user %s authenticated with backend %s", username, backend.GetName())
+			break
+		} else if getUserErr != nil && err == nil {
+			err = getUserErr
+		}
+	}
+
+	// If authenticated is true, it means at least one backend didn't fail and
+	// accepted the user. If so, honor the backend and clear the error.
+	if authenticated {
+		err = nil
+	}
+
+	return authenticated, err
+}
+
+// AuthAclCheck checks user/topic/acc authorization.
+func (b *Backends) AuthAclCheck(clientid, username, topic string, acc int) (bool, error) {
+	var aclCheck bool
+	var err error
+
+	// If prefixes are enabled, check if username has a valid prefix and use the correct backend if so.
+	// Else, check all backends.
+	if !b.checkPrefix {
+		return b.checkAcl(username, topic, clientid, acc)
+	}
+
+	validPrefix, bename := b.lookupPrefix(username)
+
+	if !validPrefix {
+		return b.checkAcl(username, topic, clientid, acc)
+	}
+
+	// If the backend is JWT and the token was prefixed, then strip the token. If the token was passed without a prefix then let it be handled in the common case.
+	if bename == jwtBackend {
+		prefix := b.getPrefixForBackend(bename)
+		username = strings.TrimPrefix(username, prefix+"_")
+	}
+	var backend = b.backends[bename]
+
+	// Short circuit checks when superusers are disabled.
+	if !b.disableSuperuser {
+		log.Debugf("Superuser check with backend %s", backend.GetName())
+		if !checkRegistered(bename, b.superuserCheckers) {
+			return false, fmt.Errorf("backend %s not registered to check superusers", bename)
```

Comment:
Ah, I see: this should simply return `false, nil`, there's no error since:

- You don't want to disable superusers globally.
- You also don't want to check superusers with a given backend, so don't register it.
- You then receive a user prefixed for that specific backend, so it should be checked only by that backend.
- What should you expect?
 
And I'd agree with you: well, you asked for that backend but explicitly said no superusers for it, so `false`. 
Did I get your comment right?

Reply:
Yes, I think `return false, nil` is the correct solution. In your described use-case, it would result in ACL check telling if access is granted or not (which is what I would expect).

Reply:
Oh, wait, this is actually flawed too! If `superuser` is not registered, we should just skip to check `acl`, not deny immediately.